test/saithrift/src/switch_sai_rpc_server.cpp

```diff
@@ -1204,36 +1207,37 @@ class switch_sai_rpcHandler : virtual public switch_sai_rpcIf {
           printf("Didn't find matching port to received name!\n");
           return SAI_NULL_OBJECT_ID;
       }
-     
-      max_port_attribute.id = SAI_SWITCH_ATTR_PORT_NUMBER;
-      switch_api->get_switch_attribute(1, &max_port_attribute);
-      max_ports = max_port_attribute.value.u32;
-      port_list_object_attribute.id = SAI_SWITCH_ATTR_PORT_LIST;
-      port_list_object_attribute.value.objlist.list = (sai_object_id_t *) malloc(sizeof(sai_object_id_t) * max_ports);
-      port_list_object_attribute.value.objlist.count = max_ports;
-      switch_api->get_switch_attribute(1, &port_list_object_attribute);
 
-      for (int i=0 ; i<max_ports ; i++){
-          port_lane_list_attribute.id = SAI_PORT_ATTR_HW_LANE_LIST;
-          port_lane_list_attribute.value.u32list.list = (uint32_t *) malloc(sizeof(uint32_t) * 4);
-          port_lane_list_attribute.value.u32list.count = 4;
-          port_api->get_port_attribute(port_list_object_attribute.value.objlist.list[i], 1, &port_lane_list_attribute);
+      uint32_t portNum = 0;
+      std::string front_port_number;
+      for (int k=0 ; k<port_name.length() ; k++) {
+        if (port_name[k] >= '0' && port_name[k] <= '9'){
+           front_port_number.push_back(port_name[k]);
+        }
+      }
+      portNum = std::stoi(front_port_number);
+      portNum--;
 
-          std::set<int> port_lanes;
-          for (int j=0 ; j<4 ; j++){
-              port_lanes.insert(port_lane_list_attribute.value.u32list.list[j]);
-          }
+
+      port_lane_list_attribute.id = SAI_PORT_ATTR_HW_LANE_LIST;
+      port_lane_list_attribute.value.u32list.list = (uint32_t *) malloc(sizeof(uint32_t) * 4);
+      port_lane_list_attribute.value.u32list.count = 4;
+      port_api->get_port_attribute(port_list_object_attribute.value.objlist.list[portNum], 1, &port_lane_list_attribute);
```

Comment:
the portNum here is the front_port_number. But the index for port_list_object_attribute.value.objlist.list is not front panel port. Maybe on cavium platform it is arranged in this way, but it is not guaranteed for other platform.

I can see why the performance of original code is not good as it needs to query the HW_LANE_LIST for every port in the function. To solve the problem, I propose to save the hw_lane_list as a cache so that you do not need query it for every port in this function.